test/unit/utils_test.js

```diff
@@ -68,6 +68,1210 @@ describe('Utils', function() {
+        source: this.clientKP1.publicKey(),
+        name: "SDF-test auth",
+        value: randomBytes(48).toString("base64"),
+      });
+
+      this.txBuilderOpts = {
+        fee: 100,
+        networkPassphrase: StellarSdk.Networks.TESTNET,
+      };
+    });
+
+    afterEach(function() {
+      this.serverKP, this.clientKP1, this.clientKP2, this.txAccount, this.opAccount, this.operation = null;
+    });
+
+    it("successfully validates server and client master key signatures in the transaction", function() {
+      const challenge = StellarSdk.Utils.buildChallengeTx(
+        this.serverKP,
+        this.clientKP1.publicKey(),
+        "SDF",
+        300,
+        StellarSdk.Networks.TESTNET,
+      );
+
+      clock.tick(200);
+
+      const transaction = new StellarSdk.Transaction(
+        challenge,
+        StellarSdk.Networks.TESTNET,
+      );
+      transaction.sign(this.clientKP1);
+
+      const signedChallenge = transaction
+        .toEnvelope()
+        .toXDR("base64")
+        .toString();
+
+      expect(
+        StellarSdk.Utils.verifyChallengeTxSigners(
+          signedChallenge,
+          this.serverKP.publicKey(),
+          StellarSdk.Networks.TESTNET,
+          [this.clientKP1.publicKey()],
+        ),
+      ).to.eql([this.clientKP1.publicKey()]);
+    });
+
+    it("throws an error if the server hasn't signed the transaction", function() {
+      const transaction = new StellarSdk.TransactionBuilder(
+        this.txAccount,
+        this.txBuilderOpts,
+      )
+        .addOperation(this.operation)
+        .setTimeout(30)
+        .build();
+
+      transaction.sign(StellarSdk.Keypair.random()); // Signing with another key pair instead of the server's
+
+      const invalidsServerSignedChallenge = transaction
+        .toEnvelope()
+        .toXDR("base64")
+        .toString();
+
+      expect(() =>
+        StellarSdk.Utils.verifyChallengeTxSigners(
+          invalidsServerSignedChallenge,
+          this.serverKP.publicKey(),
+          StellarSdk.Networks.TESTNET,
+          [this.clientKP1.publicKey()],
+        ),
+      ).to.throw(
+        StellarSdk.InvalidSep10ChallengeError,
+        "Transaction not signed by server: '" + this.serverKP.publicKey() + "'",
+      );
+    });
+
+    it("throws an error if there are no clients signing the Tx", function() {
+      const challenge = StellarSdk.Utils.buildChallengeTx(
+        this.serverKP,
+        this.clientKP1.publicKey(),
+        "SDF",
+        300,
+        StellarSdk.Networks.TESTNET,
+      );
+
+      clock.tick(200);
+
+      expect(() =>
+        StellarSdk.Utils.verifyChallengeTxSigners(
+          challenge,
+          this.serverKP.publicKey(),
+          StellarSdk.Networks.TESTNET,
+          [],
+        ),
+      ).to.throw(
+        StellarSdk.InvalidSep10ChallengeError,
+        /No verifiable client signers provided, at least one G... address must be provided/,
+      );
+    });
```

Comment:
The name/description given to this test doesn't mean to me what the test is testing. The description says there are no clients who have signed the transaction, i.e. no signatures, but the test passes an empty list of signers that results in the `No verifiable client signers provided...` error message. I think if we're testing a transaction with no signatures, we should provide a list of signers so that we get further into the code.

Reply:
That's a good point. After reviewing the tests (now) it seems like my understanding of `transaction signatures` vs `signers` was not very clear when I started writing this code.

In this specific test case it is the test name that's messed up. I'll fix that and then review the other test cases